File: problem_set_2/valid_parentheses.go
package main 

import "fmt"

type Node struct {
	item rune 
	next *Node
}

type Stack struct {
	top *Node
}

func (stack *Stack) push(bracket rune) {
	node := &Node{item: bracket, next: stack.top}

	stack.top = node
}

func (stack *Stack) pop() {
	stack.top = stack.top.next
}

func (stack Stack) print() {
	current := stack.top

	for current != nil {
		fmt.Printf("%c\n", current.item)
		current = current.next
	}

}

func (stack Stack) isEmpty() bool {
	if(stack.top == nil) {
		return true
	}

	return false
}

func isOpenBracket(bracket rune) bool {
	if bracket == '(' || bracket == '[' || bracket == '{' {
		return true
	}

	return false 
}

func getMatch(bracket rune) rune {
	switch bracket {
		case ')': return '(' 
		case ']': return '['
		case '}': return '{'
		default: return 0 
	}
}

func isValid(s string) bool {
	stack := &Stack{}

	for i := 0; i < len(s); i++ {
		current := rune(s[i])

		if isOpenBracket(current) {
			stack.push(current)
		} else {
			if stack.isEmpty() {
				return false 
			} 

			if getMatch(current) != stack.top.item {
				return false
			}

			stack.pop()
		}
	}

	return true
}

func main() {
	inputString := "([)]"
	result := isValid(inputString)
	fmt.Println(result) 
}

File: problem_set_2/ValidParentheses.java
import java.util.*; 

public class ValidParentheses {
    static boolean isOpenBracket(char bracket) {
        if(bracket == '(' || bracket == '[' || bracket == '{') {
            return true; 
        }

        return false; 
    }

    static char getMatch(char bracket) {
        switch (bracket) {
            case ')': return '(';
            case ']': return '['; 
            case '}': return '}';
            default: return 0;
        }
    }

    static boolean isValid(String s) {
        Stack<Character> stack = new Stack<>(); 

        for(int i = 0; i < s.length(); i++) {
            char current = s.charAt(i); 

            if(isOpenBracket(current)) {
                stack.push(current); 
            }
            else {
                if(stack.isEmpty()) {
                    return false; 
                }

                if(stack.peek() != getMatch(current)) {
                    return false; 
                }

                stack.pop(); 
            }
        }

        return true; 
    }

    public static void main(String[] args) {
        System.out.println(isValid("([)]")); // false
    }
}